File: product-service/src/main/java/ru/otus/spring/belov/product_service/service/ProductService.java
package ru.otus.spring.belov.product_service.service;

import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import ru.otus.spring.belov.product_service.dto.product.ProductFilter;
import ru.otus.spring.belov.product_service.dto.product.ProductItem;
import ru.otus.spring.belov.product_service.dto.product.ProductItemFull;

import java.util.List;

/**
 * Сервис по работе с продуктами
 */
public interface ProductService {

    /**
     * Возвращает продукты
     * @param productFilter запрос на фильтрацию
     * @param pageable      пагинация
     * @return продукты
     */
    Page<ProductItem> getProducts(ProductFilter productFilter, Pageable pageable);

    /**
     * Возвращает продукт
     * @param id идентификатор продукта
     * @return продукт
     */
    ProductItemFull getProductById(Long id);

    /**
     * Возвращает продукты из корзины
     * @return продукты из корзины
     */
    Page<ProductItem> getTrash(Pageable pageable);

    /**
     * Переводит продукты в корзину
     * @param ids идентификаторы продуктов
     */
    void moveToTrash(List<Long> ids);

    /**
     * Переводит продукты из корзины
     * @param ids идентификаторы продуктов
     */
    void restoreTrash(List<Long> ids);

    /**
     * Удаляет продукты
     * @param ids идентификаторы продуктов
     */
    void delete(List<Long> ids);

    /**
     * Возвращает сохранённый/изменённый продукт
     * @param saveProductRequest запрос на изменение/сохранение
     */
    ProductItemFull saveProduct(ProductItemFull saveProductRequest);
}

File: admin-ui/src/ts/services/ProductService.ts
import {OnlyInstantiableByContainer, Singleton} from "typescript-ioc";
import axios from "axios";
import {SpringPageable, TableData} from "@/ts/components/common/PageableTable";

/**
 * Сервис по работе с авторизацией
 */
@Singleton
@OnlyInstantiableByContainer
export class ProductService {

    /**
     * Возвращает продукты
     * @param filter   фильтр
     * @param pageable пагинация
     */
    async getProducts(filter: ProductFilter, pageable: SpringPageable): Promise<TableData<ProductItem>> {
        return (await axios.get<TableData<ProductItem>>("/product-service/admin/products/", {
            params: {
                ...pageable,
                ...filter
            }
        })).data;
    }

    /**
     * Возвращает продукт
     * @param id идентификатор продукта
     */
    async getProduct(id: number): Promise<ProductItem> {
        return (await axios.get<ProductItem>(`/product-service/products/${id}`)).data;
    }

    /**
     * Возвращает содержимое корзины
     */
    async getTrash(pageable: SpringPageable): Promise<TableData<ProductItem>> {
        return (await axios.get<TableData<ProductItem>>("/product-service/admin/products/trash", {
            params: pageable
        })).data;
    }

    /**
     * Переносит в корзину
     * @param ids идентификаторы продуктов
     */
    async moveToTrash(ids: number[]): Promise<void> {
        await axios.post("/product-service/admin/products/trash", ids);
    }

    /**
     * Переносит в корзину
     * @param ids идентификаторы категории
     */
    async trashRestore(ids: number[]): Promise<void> {
        await axios.post("/product-service/admin/products/trash/restore", ids);
    }

    /**
     * Удаляет с концами
     * @param ids идентификаторы категории
     */
    async delete(ids: number[]): Promise<void> {
        await axios.delete("/product-service/admin/products/", {
            data: ids
        });
    }

    /**
     * Сохраняет категорию
     * @param product сохраняемый продукт
     */
    async saveProduct(product: SaveProductRequest): Promise<ProductItem> {
        return (await axios.post<ProductItem>("/product-service/admin/products/", product)).data;
    }
}

/** Фильтр продуктов */
export type ProductFilter = {
    /** Категория */
    categoryId?: number;
    /** Признак того что продукт в корзине */
    deleted?: boolean;
    /** Признак того что опубликовано */
    published?: boolean;
}

/** Продукт */
export type ProductItem = {
    /** Идентификатор */
    id: number;
    /** Заголовок */
    title: string;
    /** Признак что ресурс в корзине */
    deleted: boolean;
    /** Опубликован */
    published: boolean;
    /** Индекс сортировки */
    sortIndex: number;
}

export type SaveProductRequest = ProductItem & {
    /** Список идентификаторов категорий */
    categories: number[];
}